spock-core/src/main/java/org/spockframework/compiler/WhereBlockRewriter.java

```diff
@@ -66,11 +69,33 @@ private void rewrite() {
         resources.getErrorReporter().error(e);
       }
 
+    validateWhereBlockVariableOrder();
+
     whereBlock.getAst().clear();
     handleFeatureParameters();
     createDataProcessorMethod();
   }
 
+  private void validateWhereBlockVariableOrder() {
+    try {
+      Parameter[] parameters = whereBlock.getParent().getAst().getParameters();
+      if (parameters.length == 0 || dataProcessorVars.size() == 0) return;
+
+      for (int i = 0; i < Math.min(parameters.length, dataProcessorVars.size()); i++)
```

Comment:
`Math.min(parameters.length, dataProcessorVars.size())` should be saved into a local var